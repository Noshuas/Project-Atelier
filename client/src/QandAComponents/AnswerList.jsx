import React from 'react';
import Answer from './Answer.jsx';

const AnswerList = (props) => {
  if (!props.moreAnswers) {
    return (
      <div>
        {props.answers.map((answer, index) => {
          if (index < 2) {
            return <Answer info={answer} key={index}/>;
          }
        })}
        <ExandAnswers handleMoreAnswers={props.handleMoreAnswers} numAnswers={props.answers.length}/>
      </div>
    );
  }
  return (
    <div className='answer-list'>
      {props.answers.map((answer, index) => {
        if (index < props.answers.length) {
          return <Answer info={answer} key={index}/>;
        }
      })}
      <a className='more-answers-lnk' href='#' onClick={props.handleMoreAnswers}>Collapse answers</a>
    </div>
  );
};

function ExandAnswers(props) {
  if (props.numAnswers === 0) {
    return <span>This question currently has no answers</span>;
  }
  if (props.numAnswers > 2) {
    return <a className='more-answers-lnk' href='#' onClick={props.handleMoreAnswers}> See more answers</a>;
  }
  return null;
}

export default AnswerList;