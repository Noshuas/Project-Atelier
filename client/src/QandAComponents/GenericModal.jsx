import React from 'react';
import ReactDOM from 'react-dom';
// could make more generic

const modal_styles = {
  position: 'fixed',
  top: '50%',
  left: '50%',
  padding: '50px',
  backgroundColor: 'white',
  zIndex: 1000,
  transform: 'translate(-50%, -50%)',
};

const form_style = {
  'max-height': 'calc(100vh - 300px)',
  'overflow-y': 'auto',
  'overflow-x': 'hidden'
};

const overlay_styles = {
  position: 'fixed',
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  backgroundColor: 'rgba(0, 0, 0, .7)',
  zIndex: 1000
};

function GenericModal ({open, children, onClose}) {
  if (!open) {
    return null;
  }
  return ReactDOM.createPortal(
    <div>
      <div style={overlay_styles}/>
      <div style={modal_styles}>
        <button className="exit-modal" onClick={onClose}>Exit</button>
        <div style={form_style}>
          {children}
        </div>
      </div>
      {/* <button onClick={onClose}>Submit</button> */}
    </div>,
    document.getElementById('portal')
  );
}

export default GenericModal;