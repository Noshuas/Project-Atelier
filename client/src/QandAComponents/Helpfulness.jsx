import React, {useState, useEffect} from 'react';
import QAapiCalls from './QandAAPIcalls.js';

function Helpfulness (props) {

  let [feedback, setFeedback] = useState(false);

  let handleClick = (event) => {
    event.preventDefault();
    QAapiCalls.postHelpfullnessFeedback(props.QorA, props.id, 'helpful')
      .then( (res) => {
        setFeedback(true);
      })
      .catch( (err) => { console.log('err:', err); });
  };
  if (!feedback) {

    return (
      <>
        <span> Helpful? <button className="link-style-btn" onClick={handleClick}>Yes</button> ({props.helpfulness})
        </span>
      </>
    );
  }
  return (
    <>
      <span> Helpful?  Yes ({props.helpfulness + 1})  </span>
    </>
  );
}

export default Helpfulness;