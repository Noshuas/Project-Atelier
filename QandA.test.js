import React, { useState } from 'react';
import renderer from 'react-test-renderer';
import Helpfulness from './client/src/QandAComponents/Helpfulness.jsx';
import GenericModal from './client/src/QandAComponents/GenericModal.jsx';
import ReactDOM from 'react-dom';
import Helpers from './client/src/QandAComponents/helpers.js';
import AnswerList from './client/src/QandAComponents/AnswerList.jsx';
import { render, screen } from '@testing-library/react';
/**
 * @jest-environment jsdom
 */
 test('renders learn react link', () => {
  render(<Helpfulness helpfulness='7' QorA='questions' id='17076' />);
  const linkElement = screen.getByText(/learn react/i);
  expect(linkElement).toBeInTheDocument();
});
// it('renders without crashing', () => {
//   const div = document.createElement("div");
//   ReactDOM.render(<Helpfulness helpfulness='7' QorA='questions' id='17076' />, div);
// });

test('Helpfulness takes the helpfulness prop and displays it correctly', () => {
  const component = renderer.create(
    <Helpfulness helpfulness='7' QorA='questions' id='17076' />
  );
  expect(tree).toMatchSnapshot();

  let tree = component.toJSON();
  let spanText = Helpers.testHelp(tree.children);
  expect(spanText).toBe(' Helpful? Yes (7)');

  const component2 = renderer.create(
    <Helpfulness helpfulness='0' QorA='questions' id='17076' />
  );
  tree = component2.toJSON();
  spanText = Helpers.testHelp(tree.children);
  expect(spanText).toBe(' Helpful? Yes (0)');

  const component3 = renderer.create(
    <Helpfulness helpfulness='####' QorA='questions' id='17076' />
  );
  tree = component3.toJSON();
  spanText = Helpers.testHelp(tree.children);
  expect(spanText).toBe(' Helpful? Yes (####)');
});

test('Generic Modal does not render when Open is false', () => {
  const testClose = () => {};
  let component = renderer.create(
    <GenericModal open={false} onClose={testClose} >
      <div>
        Hello
      </div>
    </GenericModal>
  );
  let tree = component.toJSON();
  expect(tree).toBe(null);

  // component = renderer.create(
  //   <GenericModal open={true} onClose={testClose} >
  //     Hello
  //   </GenericModal>
  // );
  // tree = component.toJSON();
  // expect(tree.children).toBe(['Hello']);
});

// Answers List Rendering
// what asnwers list nees
// answers, moreAnswer boolean, handleMoreAnswers func

// only loads two to start
test('Answer list loads two answers at the start', () => {
  let testData = [
    {
      'answer_id': 8,
      'body': 'What a great question!',
      'date': '2018-01-04T00:00:00.000Z',
      'answerer_name': 'metslover',
      'helpfulness': 8,
      'photos': [],
    },
    {
      'answer_id': 9,
      'body': 'What a great question!',
      'date': '2018-01-04T00:00:00.000Z',
      'answerer_name': 'metslover',
      'helpfulness': 8,
      'photos': [],
    },
    {
      'answer_id': 10,
      'body': 'What a great question!',
      'date': '2018-01-04T00:00:00.000Z',
      'answerer_name': 'metslover',
      'helpfulness': 8,
      'photos': [],
    }];
  const component = renderer.create(
    <AnswerList answers={testData} moreAnswers={false} handleMoreAnswers={() => {}} />
  );
  let tree = component.toJSON();
  console.log(tree);
});
// see more to load more

// does not show see more if there are none
