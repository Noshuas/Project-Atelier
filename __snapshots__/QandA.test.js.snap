// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helpfulness takes the helpfulness prop and displays it correctly 1`] = `undefined`;
